app/controllers/recipes_controller.rb

```diff
@@ -0,0 +1,100 @@
+class RecipesController < ApplicationController
+  # before_action :set_recipe, only: %i[show destroy]
+  before_action :authenticate_user!, only: %i[show toggle_visibility shopping_list]
+  # include RecipesHelper
+
+  def index
+    @recipes = Recipe.where(user_id: current_user.id)
+    # @recipes = Recipe.all
+  end
+
+  def new
+    @recipe = Recipe.new
+  end
+
+  def create
+    @recipe = Recipe.new(user_id: current_user.id, **recipe_params)
+    if @recipe.save
+      flash[:notice] = 'Recipe created successfully'
+      redirect_to recipes_path
+    else
+      flash[:alert] = 'Recipe creation failed'
+      render :new
+    end
+  end
+
+  def toggle_visibility
+    @recipe = Recipe.find(params[:id])
+    if @recipe.user == current_user
+      @recipe.public = !@recipe.public
+      @recipe.save
+    else
+      flash[:notice] = 'Only owner can update visibility'
+    end
+
+    redirect_to request.referrer
+  end
+
+  def shopping_list
+    @recipes = Recipe.where(user: current_user)
+    @foods = Food.where(user: current_user)
+
+    @missing_foods = []
+    @total_value = 0
+    @recipes.each do |recipe|
+      recipe.recipes_foods.each do |recipes_food|
+        unless @foods.include?(recipes_food.food)
+          @missing_foods << recipes_food
+          @total_value += (recipes_food.food.price * recipes_food.quantity)
+        end
+      end
+    end
+    render 'shopping_list'
+  end
```

Comment:
- [ ] You did a great job making the app function as expected. :+1: I think generating a shopping list is not working as expected. Please check the following images as a reference. 😎️

**Food**
![Screenshot from 2023-07-07 21-19-06](https://github.com/Iqbal-Elham/Recipe-app/assets/17243325/4295f18f-5198-47cf-aadf-e3f8ebdf9032)

**Recipe**
![Screenshot from 2023-07-07 21-20-09](https://github.com/Iqbal-Elham/Recipe-app/assets/17243325/35de4166-5ad3-45dc-83f6-41c7fdf04ca0)

**Shopping list**
![Screenshot from 2023-07-07 21-20-28](https://github.com/Iqbal-Elham/Recipe-app/assets/17243325/a5872613-4b23-4b12-8eff-f354e473ac60)

- I have 6kgs of apples and I need 10kgs to bake my Apple pie. The shopping list should reflect the missing 4kgs and the total price. Please address the issue and ensure that it works well when we have multiple foods and recipes. 🤝️